File: src/java/tusan/nodes/base/loops/While.java
package net.tutla.tums.tusan.nodes.base.loops;

import net.tutla.tums.tusan.Node;
import net.tutla.tums.tusan.lexer.Token;
import net.tutla.tums.tusan.lexer.TokenType;
import net.tutla.tums.tusan.nodes.Statement;
import net.tutla.tums.tusan.nodes.expression.Condition;

public class While extends Node {
    public int currentPos;
    public int endPos;
    public boolean cond;

    public While(Token token){
        super(token);
        currentPos = interpreter.pos;
        endPos = currentPos;
    }

    public While create(){
        check();

        // this method is inefficient but does it look like i fucking care
        boolean fned = false;
        int structures = 0;
        while (!fned){
            Token nxt = interpreter.nextToken();
            if (nxt.type==TokenType.ENDSTRUCTURE){
                if (structures == 0){
                    endPos = interpreter.pos;
                    fned = true;
                } else {
                    structures--;
                }
            } else if (nxt.type == TokenType.STRUCTURE) {
                structures++;
            } else {
                interpreter.nextToken();
            }
        }
        interpreter.pos = currentPos;
        check();

        while (cond){
            Token nxt = interpreter.nextToken();
            if (nxt.type==TokenType.ENDSTRUCTURE){
                if (structures == 0){
                    check();
                } else {
                    structures--;
                }
            } else if (nxt.type == TokenType.STRUCTURE) {
                structures++;
            } else {
                new Statement(nxt).create();
            }
        }
        interpreter.pos = endPos;

        return this;
    }

    public void check(){
        interpreter.pos = currentPos;
        Condition condition = new Condition(token).create();
        interpreter.expectToken(TokenType.KEYWORD, "do");
        cond = condition.value;
    }
}

File: src/python/src/tusan/nodes/base/loops/While.py
from tusk.node import Node
from tusk.token import Token
from tusk.nodes.condition import *
class WhileNode(Node):
    def __init__(self, token: Token):
        from tusk.interpreter import Interpreter
        from tusk.nodes.statement import StatementNode
        self.token = token
        self.interpreter = token.interpreter
        self.cpos = self.interpreter.pos # position of condition start
        self.ecpos = self.cpos # position at ENDSTRUCTURE

    async def create(self):
        from tusk.nodes.statement import StatementNode
        await self.check()
        
        fned = False
        internal_structure_count = 0
        while fned != True:
            nxt_tkn = self.interpreter.next_token()
            if nxt_tkn.type=="ENDSTRUCTURE":
                if internal_structure_count == 0:
                    self.ecpos = self.interpreter.pos
                    fned = True
                else:
                    internal_structure_count -= 1
            elif nxt_tkn.type == "STRUCTURE":
                internal_structure_count += 1
            else: 
                self.interpreter.next_token()
        self.interpreter.pos = self.cpos
        await self.check() # this is just to skip the condition   
        
        internal_structure_count = 0
        while self.end_done:
            nxt_tkn = self.interpreter.next_token()
            if nxt_tkn.type=="ENDSTRUCTURE":
                if internal_structure_count == 0:
                    await self.check()
                else:
                    internal_structure_count -= 1
            elif nxt_tkn.type == "STRUCTURE":
                internal_structure_count += 1
            else: 
                await StatementNode(nxt_tkn).create()

             # recheck the condition
        self.interpreter.pos = self.ecpos
        return self

    async def check(self):
        self.interpreter.pos = self.cpos
        condition = await ConditionNode(self.token).create()
        self.interpreter.expect_token("KEYWORD:do")
        self.end_done = condition.value
        return self